### DISCUSSION TITLE
ParallelForEach does not execute my Activities in parallel

### DISCUSSION THREAD

**Discussion by rudi77** [elsa-workflows/elsa-core#5431] [Q&A]
Hi everyone,

I need some help with executing my custom activity in parallel using different configurations. I'm using the ParallelForEach activity, but I've noticed that the activities are executed sequentially instead of in parallel. Could anyone provide examples or guidance on how to properly use ParallelForEach for parallel execution?

Here's my current code:

```csharp
public static ParallelForEach CreateParallelTyphonServices(List<TyphonModel> typhonModels)
{
    if (typhonModels == null)
    {
        throw new ArgumentNullException("TyphonModels is null!");
    }
     var typhonResponse = new Variable<string>();

    var parallelForEach = new ParallelForEach()
    {
        Items = new Input<object>(context => typhonModels.ToList()),
        Body = new TyphonServiceProxyActivity(new Api<TyphonBaseRequest, JObject>(), typhonResponse)
            {
                RunAsynchronously = true
            };
    };

    return parallelForEach;
}
```

And here is my custom Activity class:

```csharp
[Activity(
    "TyphonServiceProxy", 
    "Bludelta", 
    "Activity which can be used to call a typhon service via HTTP",
    Kind = ActivityKind.Job
)]
public class TyphonServiceProxyActivity : CodeActivity<string>
{
    private readonly IBcLog _logger;
    private readonly IApi<TyphonBaseRequest, JObject> _api;

    public TyphonServiceProxyActivity(IApi<TyphonBaseRequest, JObject> api, Variable<string> typhonResponse)
    {
        _logger = this.Logger(nameof(TyphonServiceProxyActivity));
        _api = api ?? throw new ArgumentNullException(nameof(api));
        Result = new(typhonResponse);
    }

    // Some Properties

    protected override async ValueTask ExecuteAsync(ActivityExecutionContext context)
    {
        await Measure.ExecuteAsync2($"{nameof(TyphonServiceProxyActivity)} with ModelName '{ModelName}' and ModelVersion '{ModelVersion}'", async () => await ExecuteInternalAsync(context), _logger);
    }

    private async ValueTask ExecuteInternalAsync(ActivityExecutionContext context)
    {
        try
        {
            // CurrentValue variable is set by the ParallelForEach activity
            var config = context.GetVariable<TyphonModel>("CurrentValue");
            if (config != null)
            {
                Name = config.Name;
                Url = config.Url;
                ModelName = config.Name;
                ModelVersion = config.Version;
                InputFormat = config.InputFormat;
                ResponseFormat = config.ResponseFormat;
            }

            // Some code

            await SendAsync(context, null, invoice, pageInfosAndImages);

            await context.CompleteActivityWithOutcomesAsync(MyOutcomes.Done);
        }
        catch (Exception ex)
        {
            await context.CompleteActivityWithOutcomesAsync(MyOutcomes.Fault);
        }
    }

    // More code
}
```

I'm curious to know if there's something I need to change in my activity to enable parallel execution. Any advice or examples would be greatly appreciated!

**Comment by glime-ai** [comment]
Based on the context provided, it seems that you are trying to execute custom activities in parallel using Elsa Workflow's `ParallelForEach` activity. However, you've noticed that the activities are being executed sequentially.

From the information provided in the GitHub issue [#4183](https://github.com/elsa-workflows/elsa-core/issues/4183), it is mentioned that the `ParallelForEach` activity in Elsa 2 does not truly execute activities in parallel. It schedules all iterations at once, but the workflow runner processes them one by one. Elsa 3 introduces the capability for activities of kind `Job` and `Task` to be executed in the background, which allows for parallel execution.

In your custom activity `TyphonServiceProxyActivity`, you have specified the `Kind` as `ActivityKind.Job`, which should allow for parallel execution in Elsa 3. However, if you are using Elsa 2, this parallelism might not be achieved as expected.

To ensure that your activities are executed in parallel, you should be using Elsa 3 and make sure that your activities are of the type that supports parallel execution (`Job` or `Task`). Additionally, you should verify that your `ParallelForEach` activity is properly configured to handle parallel execution.

If you are already using Elsa 3 and still facing issues, you might want to check the implementation of your `TyphonServiceProxyActivity` to ensure that it is correctly handling parallel execution. You may also want to review the `CurrentValue` variable usage within the `ParallelForEach` loop to avoid potential race conditions as mentioned in the GitHub issue [#5389](https://github.com/elsa-workflows/elsa-core/issues/5389).

Unfortunately, without more specific details about the version of Elsa you are using and the exact behavior you are observing, it's challenging to provide a more precise answer. If you are using Elsa 2, consider upgrading to Elsa 3 for better support for parallel execution.

If you need further assistance, please provide additional details such as the version of Elsa you are using and any specific errors or behaviors you are encountering.

This comment was generated by [Glime](https://glimelab.ai/?utm_source=dRbvTb7IhCSYiI&utm_medium=github).

**Reply by rudi77** [reply]
Here are some further details. I am using Elsa 3.1.2
